dev/PipsControl/PipsControl.cpp

```diff
@@ -0,0 +1,335 @@
+        if (button) {
+            winrt::AutomationProperties::SetName(button, ResourceAccessor::GetLocalizedStringResource(SR_PipsControlPreviousPageButtonTextName));
+            m_previousPageButtonClickRevoker = button.Click(winrt::auto_revoke, { this, &PipsControl::OnPreviousButtonClicked });
+        }
+    }(GetTemplateChildT<winrt::Button>(c_previousPageButtonName, *this));
+
+    m_nextPageButtonClickRevoker.revoke();
+    [this](const winrt::Button button) {
+        if (button) {
+            winrt::AutomationProperties::SetName(button, ResourceAccessor::GetLocalizedStringResource(SR_PipsControlNextPageButtonTextName));
+            m_nextPageButtonClickRevoker = button.Click(winrt::auto_revoke, { this, &PipsControl::OnNextButtonClicked });
+        }
+    }(GetTemplateChildT<winrt::Button>(c_nextPageButtonName, *this));
+
+    m_verticalPipsElementPreparedRevoker.revoke();
+    [this](const winrt::ItemsRepeater repeater) {
+        m_verticalPipsRepeater.set(repeater);
+        if (repeater) {
+            m_verticalPipsElementPreparedRevoker = repeater.ElementPrepared(winrt::auto_revoke, { this, &PipsControl::OnElementPrepared });
+        }
+    }(GetTemplateChildT<winrt::ItemsRepeater>(c_PipsControlPipsRepeaterName, *this));
+
+
+    m_verticalPipsScrollViewer.set(GetTemplateChildT<winrt::FxScrollViewer>(c_PipsControlScrollViewerName, *this));
+
+    m_rootGridPointerEnteredRevoker.revoke();
+    m_rootGridPointerExitedRevoker.revoke();
+
+    [this](const winrt::Grid grid) {
+        if (grid) {
+            m_rootGridPointerEnteredRevoker = grid.PointerEntered(winrt::auto_revoke, { this, &PipsControl::OnPipsControlPointerEntered });
+            m_rootGridPointerExitedRevoker = grid.PointerExited(winrt::auto_revoke, { this, &PipsControl::OnPipsControlPointerExited });
+            m_rootGridKeyDownRevoker = grid.KeyDown(winrt::auto_revoke, { this, &PipsControl::OnRootGridKeyDown });
+
+        }
+    }(GetTemplateChildT<winrt::Grid>(c_rootGridName, controlProtected));
+
+    OnNumberOfPagesChanged(0);
+    OnSelectedPageIndexChange(-1);
+    OnMaxDisplayedPagesChanged(0);
+}
+
+void PipsControl::OnPipsControlPointerEntered(winrt::IInspectable sender, winrt::PointerRoutedEventArgs args) {
+    m_isPointerOver = true;
+    UpdateNavigationButtonVisualStates();
+}
+void PipsControl::OnPipsControlPointerExited(winrt::IInspectable sender, winrt::PointerRoutedEventArgs args) {
+    // TODO : Explain WHAT'S UP and mention quarks
+    if (!isWithinBounds(args.GetCurrentPoint(*this).Position())) {
+        m_isPointerOver = false;
+        HideNavigationButtons();
+        args.Handled(true);
+    }
+
+}
+
+bool PipsControl::isWithinBounds(winrt::Point point) {
+    return point.X >= 0 && point.X <= ActualWidth() && point.Y >= 0 && point.Y <= ActualHeight();
+}
+
+void PipsControl::HideNavigationButtons() {
+    winrt::VisualStateManager::GoToState(*this, c_previousPageButtonHiddenVisualState, false);
```

Comment:
There are GoToState calls in bunch of places. Would it be simpler to have them all in a single method UpdateVisualStates and go to the correct states depending on properties in the control  like m_isPointerOver etc ?

Reply:
I agree, code as written its very hard for me to figure out what the state machine looks like

---
In reply to: [520928058](https://github.com/microsoft/microsoft-ui-xaml/pull/3592#discussion_r520928058) [](ancestors = 520928058)